server/src/main/java/org/elasticsearch/action/search/FetchSearchPhase.java

```diff
@@ -162,11 +163,18 @@ private void executeFetch(final int shardIndex, final SearchShardTarget shardTar
                               final CountedCollector<FetchSearchResult> counter,
                               final ShardFetchSearchRequest fetchSearchRequest, final QuerySearchResult querySearchResult,
                               final Transport.Connection connection) {
+        final FieldsOptionSourceAdapter adapter;
+        if (querySearchResult instanceof WrappedQuerySearchResult) {
+            adapter = ((WrappedQuerySearchResult) querySearchResult).getAdapter();
```

Comment:
`QuerySearchResult` is a transport response so we should try to avoid the wrapper. Could we create the adapter for the response dynamically here ? We don't really need to share the adapter between the query and fetch phase. I am not even sure that an interface is needed for the adapter. It could be two separate functions that are called in context (depending on the version of the connection and the search request options) ?

Reply:
> Could we create the adapter for the response dynamically here

I think in order to do that we would need the original searchSource that we use to create the adapter in the first place. The response conversion method accesses the fieldFetcher that is created inside the adapters `create`  method. The adapter also hold a reference to the `removeSourceOnResponse` flag that is initially created to re-create the `source` behaviour of the original request. This is why I think we do need to share the adapter between the phases. It holds information about the field patterns to be retrieved and how/if to restore `source`. I chose the wrapper as an easy way to transport this information here. Would you have an alternative suggestion I should look into?

Reply:
In addition, even though we seem to have the connection here, I don't see how we can access the original search request to check for the features enablement flag here. It is possible in `SearchQueryThenFetchAsyncAction` though where the adapter is created at the moment, which means we'd have to somehow transport his information here. Currently this is captured in the apdate itself.

Reply:
`SearchPhaseContext#getRequest` gives you access to the original request that contains the flag. Regarding the sharing between phases, can't we recreate the fetcher from the request ? If not, then I'd prefer that we pass the adapter explicitly in the `FetchPhase`.

Reply:
Great, that seems to work, at least looking at preliminary local test runs. I'm going to make that change and create the adapter dynamically here. Maybe this can be further improved so we don't need to create the field fetcher for every shard, I'm going to look into this.